backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/importfile/log/BankStatementImportFileLoggable.java

```diff
@@ -0,0 +1,140 @@
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import lombok.NonNull;
+import org.adempiere.service.ClientId;
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+@EqualsAndHashCode
+public class BankStatementImportFileLoggable implements ILoggable
+{
+	private static final Logger logger = LogManager.getLogger(ApiAuditLoggable.class);
+
+	private final BankStatementImportFileLogRepository bankStatementImportFileLogRepository;
+	private final BankStatementImportFileId bankStatementImportFileId;
+	private final ClientId clientId;
+	private final UserId userId;
+	private final int bufferSize;
+
+	@Nullable
+	private List<BankStatementImportFileRequestLog> buffer;
+
+	@Builder
+	public BankStatementImportFileLoggable(
+			@NonNull final BankStatementImportFileLogRepository bankStatementImportFileLogRepository,
+			@NonNull final BankStatementImportFileId bankStatementImportFileId,
+			@NonNull final ClientId clientId,
+			@NonNull final UserId userId,
+			final int bufferSize)
+	{
+		Check.assumeGreaterThanZero(bufferSize, "bufferSize");
+
+		this.bankStatementImportFileLogRepository = bankStatementImportFileLogRepository;
+		this.bankStatementImportFileId = bankStatementImportFileId;
+		this.clientId = clientId;
+		this.userId = userId;
+		this.bufferSize = bufferSize;
+	}
+
+	@Override
+	public ILoggable addLog(final String msg, final Object... msgParameters)
+	{
+		if (msg == null)
+		{
+			logger.warn("Called with msg=null; msgParameters={}; -> ignoring;", msgParameters);
+			return this;
+		}
+		final BankStatementImportFileRequestLog logEntry = createLogEntry(msg, msgParameters);
+
+		addToBuffer(logEntry);
+
+		return this;
+	}
+
+	@Override
+	public void flush()
+	{
+		final List<BankStatementImportFileRequestLog> logEntries = buffer;
+		this.buffer = null;
+
+		if (logEntries == null || logEntries.isEmpty())
+		{
+			return;
+		}
+
+		try
+		{
+			bankStatementImportFileLogRepository.insertLogs(logEntries);
+		}
+		catch (final Exception ex)
+		{
+			// make sure flush never fails
+			logger.warn("Failed saving {} log entries but IGNORED: {}", logEntries.size(), logEntries, ex);
+		}
+	}
+
+	@NonNull
+	private BankStatementImportFileRequestLog createLogEntry(@NonNull final String msg, final Object... msgParameters)
+	{
+		final LoggableWithThrowableUtil.FormattedMsgWithAdIssueId msgAndAdIssueId = LoggableWithThrowableUtil.extractMsgAndAdIssue(msg, msgParameters);
+
+		return BankStatementImportFileRequestLog.builder()
+				.message(msgAndAdIssueId.getFormattedMessage())
+				.adIssueId(msgAndAdIssueId.getAdIsueId().orElse(null))
+				.timestamp(SystemTime.asInstant())
+				.bankStatementImportFileId(bankStatementImportFileId)
+				.clientId(clientId)
+				.userId(userId)
+				.build();
+	}
+
+	private void addToBuffer(@NonNull final BankStatementImportFileRequestLog logEntry)
+	{
+		List<BankStatementImportFileRequestLog> buffer = this.buffer;
```

Comment:
why not directly use `this.buffer`?

Reply:
due to concurrency.. checkout `de.metas.async.processor.impl.WorkpackageLoggable#flush` and I believe it will make sense